include/hc.hpp

```diff
@@ -117,6 +451,188 @@ inline uint64_t get_tick_frequency() {
+                                                                                  const Kernel&);
+
+    #if __KALMAR_ACCELERATOR__ == 2 || __KALMAR_CPU__ == 2
+        template<typename Kernel, int N>
+        friend
+        completion_future launch_cpu_task_async(const std::shared_ptr<Kalmar::KalmarQueue>&,
+                                                const Kernel&,
+                                                const extent<N>&);
+    #endif
+
+    // non-tiled parallel_for_each
+    // generic version
+    template<int N, typename Kernel>
+    friend
+    completion_future parallel_for_each(const accelerator_view&,
+                                        const extent<N>&,
+                                        const Kernel&);
+
+    // 1D specialization
+    template <typename Kernel>
+    friend
+    completion_future parallel_for_each(const accelerator_view&,
+                                        const extent<1>&,
+                                        const Kernel&);
+
+    // 2D specialization
+    template <typename Kernel>
+    friend
+    completion_future parallel_for_each(const accelerator_view&,
+                                        const extent<2>&,
+                                        const Kernel&);
+
+    // 3D specialization
+    template <typename Kernel>
+    friend
+    completion_future parallel_for_each(const accelerator_view&,
+                                        const extent<3>&,
+                                        const Kernel&);
+
+    // tiled parallel_for_each, 3D version
+    template <typename Kernel>
+    friend
+    completion_future parallel_for_each(const accelerator_view&,
+                                        const tiled_extent<3>&,
+                                        const Kernel&);
+
+    // tiled parallel_for_each, 2D version
+    template <typename Kernel>
+    friend
+    completion_future parallel_for_each(const accelerator_view&,
+                                        const tiled_extent<2>&,
+                                        const Kernel&);
+
+    // tiled parallel_for_each, 1D version
+    template <typename Kernel>
+    friend
+    completion_future parallel_for_each(const accelerator_view&,
+                                        const tiled_extent<1>&,
+                                        const Kernel&);
+
+    static constexpr unsigned int max_async_ops = 5u; // TODO: retrieve directly from HSA runtime.
+
+    std::shared_ptr<Kalmar::KalmarQueue> queue_;
+
+    explicit
+    accelerator_view(std::shared_ptr<Kalmar::KalmarQueue> queue)
+        : queue_{std::move(queue)} {}
+
+    template<typename InputIterator>
+    completion_future create_blocking_marker_(InputIterator f,
+                                              InputIterator l,
+                                              memory_scope scope,
+                                              std::true_type) const
```

Comment:
Very good eyes, thank you for the observation. true_type and false_type are placeholders, in the final form I'll use proper tag dispatch (i.e. there will be a completion_future_tag and an accelerator_view_tag), which I think is a better fit here. I would absolutely emphatically not use a macro for requires / concept checking (although I love the latter to bits). In what regards type functions, those are something else that are to be added but I've kept out so as to minimise noise and because it's not very clear (to me, yet) where they should live - they are generally useful, so I would like to incorporate them in the code-base in general.